Database-Features/removing-corrupt-data.md

```diff
@@ -0,0 +1,91 @@
+spool /tmp/check_all_columns.log;
+set heading on;
+set verbose on;
+@/tmp/check_all_columns.sql;
+```
+To execute the file, run `exaplus -u sys -p <passwd> -c <connection string> -f <file name>`. 
+
+The script will run through the entire script and store the output into /tmp/check_all_column.log. Once it's finished, do the following:
+1. Look into the output and find any queries which return an error message. The table which causes an error message contains corrupt data and should be dropped.
+2. Save the DDL of the object. Many SQL Editors offer the ability to save the DDL of the object, however we also have a script to help with this. For more information, see [Create DDL for a Table](create-ddl-for-a-table.md).
+3. Drop the table. For more information, see [DROP TABLE](https://docs.exasol.com/db/latest/sql/drop_table.htm).
+4. Re-create the table using the saved DDL.
+5. Reload the data into the table from another source. 
+
+If one of the statistics tables fails, contact Exasol to delete the corresponding data. Once all of the corrupt objects are dropped, re-run the above script to make sure there are no errors. 
+
+If the database is still crashing after all of the tables are fixed, you can try to drop all of the indexes in case one of them was also corrupted. The steps for this are:
+
+1. (optional) Run the below query to generate ENFORCE INDEX statements.
+```sql
+select 'ENFORCE '||case when IS_LOCAL then 'LOCAL' else 'GLOBAL' end||' INDEX ON "'||INDEX_SCHEMA||'"."'||INDEX_TABLE||'" ('||GROUP_CONCAT(('"'||COLUMN_NAME||'"')  order by
+ordinal_position)||');' from "$EXA_INDEX_COLUMNS" where INDEX_TABLE not like 'RPL:%'  group by index_object_id, index_schema, index_table, is_local order by COUNT(*) ASC;
+```
+2. (optional) Save the results to a separate text file
+3. Run the below query to generate DROP INDEX statements
+```sql
+select 'DROP '||case when IS_LOCAL then 'LOCAL' else 'GLOBAL' end||' INDEX ON "'||INDEX_SCHEMA||'"."'||INDEX_TABLE||'" ('||GROUP_CONCAT(('"'||COLUMN_NAME||'"')  order by
+ordinal_position)||');' from "$EXA_INDEX_COLUMNS" where INDEX_TABLE not like 'RPL:%'  group by index_object_id, index_schema, index_table, is_local order by COUNT(*) ASC;
+```
+4. Save the results to a separate text file
+5. Run the generated queries which perform all of the DROP INDEX statements
+6. (optional) Run the generated queries which perform all of the ENFORCE INDEX statements.
+
+Note - the database will automatically create indexes as needed, therefore the ENFORCE INDEX statements are purely optional and should only be run after you verify that the database is behaving normal. 
+
+**If after all of the above actions, the database is still restarting during query execution, then this may point to the fact that an underlying data structure was corrupted and the database is unrecoverable. In this case, the only option would be to delete the database, re-create it, and reload all of the data.**
```

Comment:
I would suggest to remove this "recreate db" part completely. This has never happened so far and I doubt its necessity. It is just frightening everybody.

From options point of view: You could also dump all your tables into CSV files. The (remaining) data is accessible. Jumping too quickly into dropping scenario could lead to data losses.

Reply:
See my new text